fastapi/dependencies/utils.py

```diff
@@ -535,90 +535,188 @@ async def solve_dependencies(
             solved = await run_in_threadpool(call, **sub_values)
         if sub_dependant.name is not None:
             values[sub_dependant.name] = solved
-        if sub_dependant.cache_key not in dependency_cache:
-            dependency_cache[sub_dependant.cache_key] = solved
-    path_values, path_errors = request_params_to_args(
-        dependant.path_params, request.path_params
-    )
-    query_values, query_errors = request_params_to_args(
-        dependant.query_params, request.query_params
-    )
-    header_values, header_errors = request_params_to_args(
-        dependant.header_params, request.headers
-    )
-    cookie_values, cookie_errors = request_params_to_args(
-        dependant.cookie_params, request.cookies
-    )
-    values.update(path_values)
-    values.update(query_values)
-    values.update(header_values)
-    values.update(cookie_values)
-    errors += path_errors + query_errors + header_errors + cookie_errors
+        context.dependency_cache[cache_key] = solved
+
+    # solve built-in dependencies
+    if dependant.dependency_getters:
+        context.values = values
+        context.errors = errors
+        for dependency_getter in dependant.dependency_getters:
+            if inspect.iscoroutinefunction(dependency_getter):
+                await dependency_getter(context)
+            else:
+                dependency_getter(context)
+
+    return values, errors
+
+
+def get_dependent_dependency_getters(
+    dependant: Dependant,
+) -> Optional[List[DependencyGetter]]:
+    """
+    Process built-in dependencies into a list of getter functions
+    """
+    getters: List[DependencyGetter] = []
+    field: ModelField
+    for field in dependant.path_params:
+
+        def get_path_param_getter(field: ModelField) -> DependencyGetter:
+            def get_param(context: DependencySolverContext) -> None:
```

Comment:
Why are all these getters defined inline? As they don't seem to be using any local state, could you move them out and declare them at the top level?

Note that this is specially necessary for getters defined in a loop as in each iteration it will create a new function, generating more memory garbage.

Reply:
It's a style decision.  My original implementation attempted to not use the outer function at all (e.g. `get_path_param_getter()`, but crate the closure directly (e.g. `get_param()`, but that doesn't work in a loop neccessitating the outer helper function.  I kept it at the same place as the closure to keep all the logic in the same place.

You need not worry about "memory garbage".  The outer function is assigned to a local variable and gets destroyed when the "get_dependent_dependency_getters()" exits.  Each iteration just redefines the function.  But I can remove it out of the loop if you want.  It will just reduce the locality of the logic a bit.  Python is nice like that, local functions definitions are just fine.

This also happens during the setup of the endpoints so it is not critical performance-wise.

But I'm happy to move them out to the top level if you think that style is better.

Reply:
I could also refactor this to use `functools.partial()` instead of closures.